File: src/main/java/org/wysko/midis2jam2/instrument/pipe/Ocarina.java
package org.wysko.midis2jam2.instrument.pipe;

import com.jme3.math.Quaternion;
import com.jme3.scene.Spatial;
import org.wysko.midis2jam2.Midis2jam2;
import org.wysko.midis2jam2.instrument.HandedClone;
import org.wysko.midis2jam2.midi.MidiChannelSpecificEvent;

import java.util.HashMap;
import java.util.List;

import static org.wysko.midis2jam2.Midis2jam2.rad;

public class Ocarina extends HandedInstrument {
	
	/**
	 * Constructs an ocarina.
	 *
	 * @param context context to midis2jam2
	 */
	public Ocarina(Midis2jam2 context, List<MidiChannelSpecificEvent> events) throws ReflectiveOperationException {
		super(context,
				events,
				OcarinaClone.class,
				new OcarinaHandGenerator()
		);
		
		groupOfPolyphony.setLocalTranslation(32, 47, 30);
		groupOfPolyphony.setLocalRotation(new Quaternion().fromAngles(0, rad(135), 0));
		
	}
	
	static class OcarinaHandGenerator extends HashMap<Integer, HandedClone.Hands> {
		@Override
		public HandedClone.Hands get(Object key) {
			return new HandedClone.Hands(0, ((int) key + 3) % 12);
		}
	}
	
	public class OcarinaClone extends HandedClone {
		public OcarinaClone() {
			super(Ocarina.this, 0);
			Spatial ocarina = context.loadModel("Ocarina.obj", "Ocarina.bmp");
			animNode.attachChild(ocarina);
			highestLevel.attachChild(animNode);
			loadHands();
			for (int i = 0; i < rightHands.length; i++) {
				if (i == 0) rightHands[i].setCullHint(Spatial.CullHint.Dynamic);
				else rightHands[i].setCullHint(Spatial.CullHint.Always);
			}
			highestLevel.setLocalTranslation(0, 0, 18);
		}
		
		@Override
		protected void loadHands() {
			rightHands = new Spatial[12];
			for (int i = 0; i < 12; i++) {
				rightHands[i] = context.loadModel("OcarinaHand" + i + ".obj", "hands.bmp", Midis2jam2.MatType.UNSHADED, 0.9f);
			}
			for (Spatial rightHand : rightHands) {
				rightHandNode.attachChild(rightHand);
			}
		}
		
		@Override
		public void tick(double time, float delta) {
			super.tick(time, delta);
			/* Collect note periods to execute */
			if (isPlaying()) {
				assert currentNotePeriod != null;
				animNode.setLocalTranslation(0,
						0, 3 * (float) ((currentNotePeriod.endTime - time) / currentNotePeriod.duration()));
			}
		}
		
		@Override
		protected void moveForPolyphony() {
			offsetNode.setLocalRotation(new Quaternion().fromAngles(0, rad(17 * indexForMoving()), 0));
		}
	}
}

File: src/main/java/org/wysko/midis2jam2/instrument/family/pipe/Ocarina.kt
/*
 * Copyright (C) 2022 Jacob Wysko
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see https://www.gnu.org/licenses/.
 */
package org.wysko.midis2jam2.instrument.family.pipe

import com.jme3.math.Quaternion
import com.jme3.scene.Spatial
import org.wysko.midis2jam2.Midis2jam2
import org.wysko.midis2jam2.instrument.algorithmic.HandPositionFingeringManager
import org.wysko.midis2jam2.instrument.clone.HandedClone
import org.wysko.midis2jam2.midi.MidiChannelSpecificEvent
import org.wysko.midis2jam2.util.Utils.rad

/** The Ocarina. */
class Ocarina(context: Midis2jam2, events: List<MidiChannelSpecificEvent>) :
    HandedInstrument(context, events, OcarinaClone::class.java, OcarinaHandGenerator()) {

    /**
     * The ocarina hand positions are from 0 to 11 and wrap around the octave. So this is easily calculable and doesn't
     * need to be stored in XML.
     */
    internal class OcarinaHandGenerator : HandPositionFingeringManager() {
        override fun fingering(midiNote: Int): Hands {
            return Hands(0, (midiNote + 3) % 12)
        }
    }

    /** A single ocarina. */
    inner class OcarinaClone : HandedClone(this@Ocarina, 0f) {

        override val leftHands: Array<Spatial> = emptyArray()
        override val rightHands: Array<Spatial> = Array(12) {
            context.loadModel("OcarinaHand$it.obj", "hands.bmp")
        }

        override fun tick(time: Double, delta: Float) {
            super.tick(time, delta)
            /* Collect note periods to execute */
            if (isPlaying) {
                currentNotePeriod?.let {
                    animNode.setLocalTranslation(
                        0f,
                        0f,
                        3 * ((it.endTime - time) / it.duration()).toFloat()
                    )
                }
            }
        }

        override fun moveForPolyphony() {
            offsetNode.localRotation = Quaternion().fromAngles(0f, rad((17f * indexForMoving()).toDouble()), 0f)
        }

        init {
            loadHands()
            animNode.attachChild(context.loadModel("Ocarina.obj", "Ocarina.bmp"))
            highestLevel.attachChild(animNode)
            highestLevel.setLocalTranslation(0f, 0f, 18f)
        }
    }

    init {
        groupOfPolyphony.setLocalTranslation(32f, 47f, 30f)
        groupOfPolyphony.localRotation = Quaternion().fromAngles(0f, rad(135.0), 0f)
    }
}